File: binary_search.hs
--Arrays are more appropriate here, because indexing an array is O(1)
--Updating an array is O(n), but this is search, so no updates.

--import Data.Array;

main = do
    --let mylist = array (0,10) [(i,i) | i <- [4,7,9,12,16,17,22,56,59,77,93]]
    let thelist = [4,7,9,12,16,17,22,56,59,77,93]
    print (bin_search 16 thelist 0 10)
    
bin_search val lst lo hi 
    | hi<lo = -1 
    | val<pivot = bin_search val lst lo (mid-1) 
    | val>pivot = bin_search val lst (mid+1) hi 
    | val==pivot = mid 
    where 
        mid = (lo + hi) `div` 2 
        pivot = lst!!mid

File: binary_search.py
#Binary search iterative approach

def bin_search(val,lst):
    lo=0
    hi = len(lst)-1
    while True:
        mid = (hi + lo) // 2
        if lst[mid]==val:
            return mid
        elif val<lst[mid]:
            hi = mid - 1
        elif val>lst[mid]:
            lo = mid + 1
        if hi<lo:
            return -1

thelist = [4,7,9,12,16,17,22,56,59,77,93]
print bin_search(16,thelist)